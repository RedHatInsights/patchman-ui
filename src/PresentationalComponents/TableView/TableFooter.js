import { Pagination, PaginationVariant } from '@patternfly/react-core/dist/js/components/Pagination';
import { TableToolbar } from '@redhat-cloud-services/frontend-components/components/cjs/TableToolbar';
import PropTypes from 'prop-types';
import React from 'react';

const TableFooter = ({ page, perPage, onSetPage, totalItems, onPerPageSelect, paginationOUIA }) => {
    return (
        <TableToolbar>
            <Pagination
                itemCount={totalItems}
                perPage={perPage}
                page={page}
                onSetPage={onSetPage}
                onPerPageSelect={onPerPageSelect}
                widgetId={`pagination-options-menu-bottom`}
                variant={PaginationVariant.bottom}
                ouiaId={paginationOUIA}
            />
        </TableToolbar>
    );
};

TableFooter.propTypes = {
    onSetPage: PropTypes.func,
    onPerPageSelect: PropTypes.func,
    page: PropTypes.number,
    perPage: PropTypes.number,
    totalItems: PropTypes.number,
    paginationOUIA: PropTypes.string
};

export default TableFooter;
