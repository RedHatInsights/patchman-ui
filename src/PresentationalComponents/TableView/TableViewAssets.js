import { cellWidth, expandable, sortable } from '@patternfly/react-table/dist/js';
import messages from '../../Messages';
import { intl } from '../../Utilities/IntlProvider';

export const advisoriesColumns = [
    {
        title: intl.formatMessage(messages.labelsColumnsName),
        cellFormatters: [expandable],
        transforms: [sortable, cellWidth(15)],
        key: 'id'
    },
    {
        title: intl.formatMessage(messages.labelsColumnsSynopsis),
        transforms: [sortable],
        key: 'synopsis'
    },
    {
        title: intl.formatMessage(messages.labelsColumnsType),
        transforms: [sortable, cellWidth(10)],
        key: 'advisory_type'
    },
    {
        title: intl.formatMessage(messages.labelsColumnsApplicableSystems),
        transforms: [sortable, cellWidth(15)],
        key: 'applicable_systems'
    },
    {
        title: intl.formatMessage(messages.labelsColumnsPublishDate),
        transforms: [sortable, cellWidth(15)],
        key: 'public_date'
    }
];

export const systemAdvisoriesColumns = [
    {
        title: intl.formatMessage(messages.labelsColumnsName),
        cellFormatters: [expandable],
        transforms: [sortable, cellWidth(15)],
        key: 'id'
    },
    {
        title: 'Synopsis',
        transforms: [sortable, cellWidth(45)],
        key: 'synopsis'
    },
    {
        title: intl.formatMessage(messages.labelsColumnsType),
        transforms: [sortable, cellWidth(15)],
        key: 'advisory_type'
    },
    {
        title: intl.formatMessage(messages.labelsColumnsPublishDate),
        transforms: [sortable, cellWidth(25)],
        key: 'public_date'
    }
];

export const systemPackagesColumns = [
    {
        title: intl.formatMessage(messages.labelsColumnsPackage),
        transforms: [sortable, cellWidth(20)],
        key: 'name'
    },
    {
        title: intl.formatMessage(messages.labelsColumnsInstalledVersion),
        transforms: [sortable, cellWidth(15)],
        key: 'evra'
    },
    {
        title: intl.formatMessage(messages.labelsColumnsLatestVersion),
        transforms: [cellWidth(15)],
        key: 'latest_update'
    },
    {
        title: intl.formatMessage(messages.labelsColumnsStatus),
        transforms: [cellWidth(10)],
        key: 'updatable'
    },
    {
        title: intl.formatMessage(messages.labelsColumnsSummary),
        transforms: [sortable, cellWidth(40)],
        key: 'summary'
    }
];

export const packagesColumns = [
    {
        key: 'name',
        title: intl.formatMessage(messages.labelsColumnsPackage),
        transforms: [sortable, cellWidth(25)]
    },
    {
        key: 'systems_installed',
        title: intl.formatMessage(messages.labelsColumnsApplicableSystems),
        transforms: [sortable, cellWidth(15)],
        props: {
            width: 10
        }
    },
    {
        key: 'systems_updatable',
        title: intl.formatMessage(messages.labelsColumnsUpgradable),
        transforms: [sortable, cellWidth(10)],
        props: {
            width: 10
        }
    },
    {
        key: 'summary',
        title: intl.formatMessage(messages.labelsColumnsSummary),
        transforms: [sortable, cellWidth(40)],
        props: {
            width: 30
        }
    }
];

export const cvesTableColumns = [
    {
        title: intl.formatMessage(messages.labelsColumnsCveID),
        transforms: [sortable, cellWidth(100)],
        key: 'id'
    }
];
