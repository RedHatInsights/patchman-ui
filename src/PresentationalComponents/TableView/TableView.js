import { Button } from '@patternfly/react-core/dist/js/components/Button/Button';
import AnsibeTowerIcon from '@patternfly/react-icons/dist/js/icons/ansibeTower-icon'; // PF typo
import { Table, TableBody, TableHeader } from '@patternfly/react-table/dist/js/components/Table';
import { PrimaryToolbar } from '@redhat-cloud-services/frontend-components/components/cjs/PrimaryToolbar';
import { SkeletonTable } from '@redhat-cloud-services/frontend-components/components/cjs/SkeletonTable';
import globalBackgroundColor100 from '@patternfly/react-tokens/dist/js/global_BackgroundColor_100';
import PropTypes from 'prop-types';
import React from 'react';
import RemediationModal from '../../SmartComponents/Remediation/RemediationModal';
import { STATUS_LOADING, STATUS_RESOLVED } from '../../Utilities/constants';
import { arrayFromObj, buildFilterChips, convertLimitOffset } from '../../Utilities/Helpers';
import { useRemoveFilter } from '../../Utilities/Hooks';
import TableFooter from './TableFooter';

const TableView = ({
    columns,
    store: {
        rows,
        metadata,
        status,
        queryParams: { filter, search }
    },
    onCollapse,
    onSelect,
    onSetPage,
    onPerPageSelect,
    onSort,
    onExport,
    filterConfig,
    sortBy,
    remediationProvider,
    selectedRows,
    apply
}) => {
    const [
        RemediationModalCmp,
        setRemediationModalCmp
    ] = React.useState(() => () => null);
    const [page, perPage] = React.useMemo(
        () => convertLimitOffset(metadata.limit, metadata.offset),
        [metadata.limit, metadata.offset]
    );

    const showRemediationModal = data => {
        setRemediationModalCmp(() => () => <RemediationModal data={data} />);
    };

    const removeFilter = useRemoveFilter(filter, apply);
    const selectedCount = selectedRows && arrayFromObj(selectedRows).length;
    return (
        <React.Fragment>
            <PrimaryToolbar
                pagination={{
                    itemCount: metadata.total_items,
                    page,
                    perPage,
                    isCompact: true,
                    onSetPage,
                    onPerPageSelect
                }}
                filterConfig={filterConfig}
                activeFiltersConfig={{
                    filters: buildFilterChips(filter, search),
                    onDelete: removeFilter
                }}
                actionsConfig={{ actions: [remediationProvider && (
                    <React.Fragment>
                        <Button
                            isDisabled={selectedCount === 0}
                            className={'remediationButtonPatch'}
                            onClick={() =>
                                showRemediationModal(remediationProvider())
                            }
                        >
                            <AnsibeTowerIcon color={globalBackgroundColor100.value} />&nbsp;Remediate
                        </Button>
                        <RemediationModalCmp />
                    </React.Fragment>
                )] }}
                exportConfig={{ onSelect: onExport }}
                bulkSelect={onSelect && {
                    count: selectedCount,
                    items: [{
                        title: `Select none (0)`,
                        onClick: () => {
                            onSelect('none');
                        }
                    }, {
                        title: `Select page (${rows.length / 2})`,
                        onClick: () => {
                            onSelect('page');
                        }
                    },
                    {
                        title: `Select all (${metadata.total_items})`,
                        onClick: () => {
                            onSelect('all');
                        }
                    }],
                    onSelect: (value) => {
                        value ? onSelect('all') : onSelect('none');
                    },
                    checked: selectedCount === metadata.total_items ? true : selectedCount === 0 ? false : null
                }}

            />

            {status === STATUS_LOADING && <SkeletonTable colSize={5} rowSize={20} />}
            {status === STATUS_RESOLVED && (
                <React.Fragment>
                    <Table
                        aria-label="Patch table view"
                        cells={metadata.total_items && columns || ['']}
                        onSelect={onSelect}
                        rows={rows}
                        onCollapse={metadata.total_items && onCollapse}
                        canSelectAll={false}
                        onSort={onSort}
                        sortBy={metadata.total_items && sortBy}
                    >
                        <TableHeader />
                        <TableBody />
                    </Table>
                    <TableFooter
                        totalItems={metadata.total_items}
                        perPage={perPage}
                        page={page}
                        onSetPage={onSetPage}
                        onPerPageSelect={onPerPageSelect}
                    />
                </React.Fragment>)}
        </React.Fragment>
    );
};

TableView.propTypes = {
    columns: PropTypes.array,
    onCollapse: PropTypes.func,
    onSelect: PropTypes.func,
    onSetPage: PropTypes.func,
    onPerPageSelect: PropTypes.func,
    onSort: PropTypes.func,
    onExport: PropTypes.func,
    remediationProvider: PropTypes.func,
    selectedRows: PropTypes.object,
    apply: PropTypes.func,
    sortBy: PropTypes.object,
    filterConfig: PropTypes.object,
    store: PropTypes.object
};

export default TableView;
