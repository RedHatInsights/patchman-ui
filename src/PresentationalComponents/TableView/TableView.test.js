import TableView from './TableView';
import toJson from 'enzyme-to-json';
/* eslint-disable */
const testObj = {
    columns: [],
    store: {
        rows: [],
        metadata: {},
        status: false,
        queryParams: {}
    },
    onCollapse: jest.fn(),
    onSelect: jest.fn(),
    onSetPage: jest.fn(),
    onPerPageSelect: jest.fn(),
    onSort: jest.fn(),
    onExport: jest.fn(),
    filterConfig: {
        items: [{
            label: 'First filter'
        }, {
            label: 'Second filter',
            type: 'checkbox',
            filterValues: {
                items: [{ label: 'Some checkbox' }]
            }
        }]
    },
    sortBy: {},
    remediationProvider: jest.fn(),
    selectedRows: {},
    apply: jest.fn()
};

describe('TableView', () => {
    it('TableView', () => {
        const wrapper = shallow(<TableView {...testObj} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    describe('test table props', () => {
        const wrapper = shallow(<TableView {...testObj}  store = {{
            rows: [],
            metadata: { total_items: 10 },
            status: 'resolved',
            queryParams: {}
        }} />);

        it('TableView', () => {
            expect(toJson(wrapper)).toMatchSnapshot();
        });

        it('Should call onCollapse', () => {    
            const { cells }  = wrapper.find('Table').props();
            wrapper.find('Table').props().onSelect();
            expect(testObj.onSelect).toHaveBeenCalled();
            expect(cells).toEqual([]);
        });

    })



    it('TableView', () => {
        const wrapper = shallow(<TableView {...testObj}  store = {{
            rows: [],
            metadata: {},
            status: 'loading',
            queryParams: {}
        }} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    
    it('Should open remediation modal', () => {
        const wrapper = shallow(<TableView {...testObj} />);
        const { actionsConfig: { actions }}  = wrapper.find('PrimaryToolbar').props();
        const remediation = actions[0].props.children;
        remediation[0].props.onClick();
        wrapper.update();
        expect(remediation[1]).toBeTruthy();
    });

    it('Should unselect', () => {
        const wrapper = shallow(<TableView {...testObj} />);
        const { bulkSelect: { items }}  = wrapper.find('PrimaryToolbar').props();
        items[0].onClick();
        expect(testObj.onSelect).toHaveBeenCalledWith('none');
    });

    it('Should select page', () => {
        const wrapper = shallow(<TableView {...testObj} />);
        const { bulkSelect: { items }}  = wrapper.find('PrimaryToolbar').props();
        items[1].onClick();
        expect(testObj.onSelect).toHaveBeenCalledWith('page');
    });

    it('Should select all', () => {
        const wrapper = shallow(<TableView {...testObj} />);
        const { bulkSelect: { items }}  = wrapper.find('PrimaryToolbar').props();
        items[2].onClick();
        expect(testObj.onSelect).toHaveBeenCalledWith('all');
    });

    it('Should unselect all ', () => {
        const wrapper = shallow(<TableView {...testObj} />);
        const { bulkSelect: { onSelect }}  = wrapper.find('PrimaryToolbar').props();
        onSelect(false);
        expect(testObj.onSelect).toHaveBeenCalledWith('none');
    });
});
/* eslint-enable */
