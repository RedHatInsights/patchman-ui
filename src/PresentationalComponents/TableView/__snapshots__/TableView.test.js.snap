// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableView TableView 1`] = `
<Fragment>
  <PrimaryToolbar
    actionsConfig={
      Object {
        "actions": Array [
          <React.Fragment>
            <Button
              className="remediationButtonPatch"
              isDisabled={true}
              onClick={[Function]}
            >
              <AnsibeTowerIcon
                color="#fff"
                noVerticalAlign={false}
                size="sm"
              />
               Remediate
            </Button>
            <Unknown />
          </React.Fragment>,
        ],
      }
    }
    activeFiltersConfig={
      Object {
        "filters": Array [],
        "onDelete": [Function],
      }
    }
    bulkSelect={
      Object {
        "checked": false,
        "count": 0,
        "items": Array [
          Object {
            "onClick": [Function],
            "title": "Select none (0)",
          },
          Object {
            "onClick": [Function],
            "title": "Select page (0)",
          },
          Object {
            "onClick": [Function],
            "title": "Select all ()",
          },
        ],
        "onSelect": [Function],
        "toggleProps": Object {
          "data-ouia-component-type": "bulk-select-toggle-button",
        },
      }
    }
    exportConfig={
      Object {
        "onSelect": [MockFunction],
      }
    }
    filterConfig={
      Object {
        "items": Array [
          Object {
            "label": "First filter",
          },
          Object {
            "filterValues": Object {
              "items": Array [
                Object {
                  "label": "Some checkbox",
                },
              ],
            },
            "label": "Second filter",
            "type": "checkbox",
          },
        ],
      }
    }
    pagination={
      Object {
        "isCompact": true,
        "itemCount": undefined,
        "onPerPageSelect": [MockFunction],
        "onSetPage": [MockFunction],
        "ouiaId": "top-undefined",
        "page": NaN,
        "perPage": undefined,
      }
    }
    toggleIsExpanded={[Function]}
  />
</Fragment>
`;

exports[`TableView TableView 2`] = `
<Fragment>
  <PrimaryToolbar
    actionsConfig={
      Object {
        "actions": Array [
          <React.Fragment>
            <Button
              className="remediationButtonPatch"
              isDisabled={true}
              onClick={[Function]}
            >
              <AnsibeTowerIcon
                color="#fff"
                noVerticalAlign={false}
                size="sm"
              />
               Remediate
            </Button>
            <Unknown />
          </React.Fragment>,
        ],
      }
    }
    activeFiltersConfig={
      Object {
        "filters": Array [],
        "onDelete": [Function],
      }
    }
    bulkSelect={
      Object {
        "checked": false,
        "count": 0,
        "items": Array [
          Object {
            "onClick": [Function],
            "title": "Select none (0)",
          },
          Object {
            "onClick": [Function],
            "title": "Select page (0)",
          },
          Object {
            "onClick": [Function],
            "title": "Select all ()",
          },
        ],
        "onSelect": [Function],
        "toggleProps": Object {
          "data-ouia-component-type": "bulk-select-toggle-button",
        },
      }
    }
    exportConfig={
      Object {
        "onSelect": [MockFunction],
      }
    }
    filterConfig={
      Object {
        "items": Array [
          Object {
            "label": "First filter",
          },
          Object {
            "filterValues": Object {
              "items": Array [
                Object {
                  "label": "Some checkbox",
                },
              ],
            },
            "label": "Second filter",
            "type": "checkbox",
          },
        ],
      }
    }
    pagination={
      Object {
        "isCompact": true,
        "itemCount": undefined,
        "onPerPageSelect": [MockFunction],
        "onSetPage": [MockFunction],
        "ouiaId": "top-undefined",
        "page": NaN,
        "perPage": undefined,
      }
    }
    toggleIsExpanded={[Function]}
  />
  <SkeletonTable
    canSelectAll={false}
    colSize={5}
    hasRadio={false}
    isSelectable={false}
    paddingColumnSize={0}
    rowSize={20}
  />
</Fragment>
`;

exports[`TableView test table props TableView 1`] = `
<Fragment>
  <PrimaryToolbar
    actionsConfig={
      Object {
        "actions": Array [
          <React.Fragment>
            <Button
              className="remediationButtonPatch"
              isDisabled={true}
              onClick={[Function]}
            >
              <AnsibeTowerIcon
                color="#fff"
                noVerticalAlign={false}
                size="sm"
              />
               Remediate
            </Button>
            <Unknown />
          </React.Fragment>,
        ],
      }
    }
    activeFiltersConfig={
      Object {
        "filters": Array [],
        "onDelete": [Function],
      }
    }
    bulkSelect={
      Object {
        "checked": false,
        "count": 0,
        "items": Array [
          Object {
            "onClick": [Function],
            "title": "Select none (0)",
          },
          Object {
            "onClick": [Function],
            "title": "Select page (0)",
          },
          Object {
            "onClick": [Function],
            "title": "Select all (10)",
          },
        ],
        "onSelect": [Function],
        "toggleProps": Object {
          "data-ouia-component-type": "bulk-select-toggle-button",
        },
      }
    }
    exportConfig={
      Object {
        "onSelect": [MockFunction],
      }
    }
    filterConfig={
      Object {
        "items": Array [
          Object {
            "label": "First filter",
          },
          Object {
            "filterValues": Object {
              "items": Array [
                Object {
                  "label": "Some checkbox",
                },
              ],
            },
            "label": "Second filter",
            "type": "checkbox",
          },
        ],
      }
    }
    pagination={
      Object {
        "isCompact": true,
        "itemCount": 10,
        "onPerPageSelect": [MockFunction],
        "onSetPage": [MockFunction],
        "ouiaId": "top-undefined",
        "page": NaN,
        "perPage": undefined,
      }
    }
    toggleIsExpanded={[Function]}
  />
  <Table
    aria-label="Patch table view"
    borders={true}
    canSelectAll={false}
    canSortFavorites={true}
    cells={Array []}
    className=""
    contentId="expanded-content"
    dropdownDirection="down"
    dropdownPosition="right"
    expandId="expandable-toggle"
    gridBreakPoint="grid-md"
    isStickyHeader={false}
    onCollapse={[MockFunction]}
    onSelect={[MockFunction]}
    onSort={[MockFunction]}
    ouiaSafe={true}
    role="grid"
    rowLabeledBy="simple-node"
    rows={Array []}
    selectVariant="checkbox"
    sortBy={Object {}}
    variant={null}
  >
    <TableHeader />
    <Component />
  </Table>
  <TableFooter
    onPerPageSelect={[MockFunction]}
    onSetPage={[MockFunction]}
    page={NaN}
    paginationOUIA="bottom-undefined"
    totalItems={10}
  />
</Fragment>
`;
